### PR TITLE
Enable to monitor batch request failure and retries

### PR DESCRIPTION
**Pull request by simukappu**
*Issue #, if available:*
#150

*Description of changes:*
* Add *drop_failed_records_after_batch_request_retries* config param to enable you to return failed chunk to Fluentd buffer for retrying
* Increment *num_errors* to monitor batch request failure from "monitor_agent" or "fluent-plugin-prometheus" when *drop_failed_records_after_batch_request_retries* is true (default)
* Add *monitor_num_of_batch_request_retries* config param to enable to increment *num_errors* for each plugin internal retries to monitor batch request retries from "monitor_agent" or "fluent-plugin-prometheus"
* Including #162

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: lib/fluent/plugin/kinesis_helper/api.rb [hunk 1]
```diff
@@ -164,6 +177,49 @@ def give_up_retries(failed_records)
...
+                target_failed_record = service_unavailable_exception_records.first || failed_records.first
+                target_error = service_unavailable_exception_records.empty? ?
+                  Aws::Firehose::Errors::ServiceError :
+                  Aws::Firehose::Errors::ServiceUnavailableException
+              end
+              log.error("Raise #{target_failed_record[:error_code]} and return chunk to Fluentd buffer for retrying")
+              raise target_error.new(Seahorse::Client::RequestContext.new, target_failed_record[:error_message])
+            else
+              # Increment num_errors to monitor batch request failure from "monitor_agent" or "fluent-plugin-prometheus"
+              increment_num_errors
```
Comment (jooohn) [inline lib/fluent/plugin/kinesis_helper/api.rb:207]: I may not understand this intention accurately, but I have some questions.

1. Is it ok to call `increment_num_errors` regardless of `@monitor_num_of_batch_request_retries`?
2. Is it ok not to call `increment_num_errors` in the `if-true` case of L181?

It's a bit hard to follow when `increment_num_errors` is called and when it's not.

It would be straightforward if we can jut add the following line at L96 (in the new commit) instead of everything, but I'm not sure if it's appropriate.
```ruby
increment_num_errors if @monitor_num_of_batch_request_retries
```

Comment (simukappu) [reply lib/fluent/plugin/kinesis_helper/api.rb:207]: Thanks for your comments! Yes, it would work as I expected. This enhancement will work as follows:

### current behavior
* No metrics will be updated by batch request retries due to Kinesis over capacity or any other causes
* Give up retries after retrying request retries as *retries_on_batch_request*, and will NOT return records to Fluentd buffer (= will NOT retry by Fluentd retry mechanism)
* No metrics will be updated when giving up request retries (= CANNOT monitor anything from monitor_agent)

### implemented behavior: default
* No metrics will be updated by batch request retries due to Kinesis over capacity or any other causes
* Give up retries after retrying request retries as *retries_on_batch_request*, and will NOT return records to Fluentd buffer (= will NOT retry by Fluentd retry mechanism)
* *num_errors* metrics will be updated by this plugin itself when giving up request retries (= can monitor from monitor_agent)

### implemented behavior: raise_error_on_batch_request_failure = true
* No metrics will be updated by batch request retries due to Kinesis over capacity or any other causes
* Give up retries after retrying request retries as *retries_on_batch_request*, and will return records to Fluentd buffer (= will retry by Fluentd retry mechanism)
* *num_errors* metrics will be updated by Fluentd retry mechanism when giving up request retries (= can monitor from monitor_agent)

### implemented behavior: monitor_num_of_batch_request_retries = true
* *num_errors* metrics will be updated by batch request retries due to Kinesis over capacity or any other causes (= can monitor from monitor_agent)
* Give up retries after retrying request retries as *retries_on_batch_request*, and will NOT return records to Fluentd buffer (= will NOT retry by Fluentd retry mechanism)
* *num_errors* metrics will be updated by this plugin itself when giving up request retries (= can monitor from monitor_agent)

Comment (jooohn) [reply lib/fluent/plugin/kinesis_helper/api.rb:207]: Thanks for explaining! Now I understand the expected behavior and I think the current implementation will work as expected. I didn't know fluentd will retry when error raised, and increase `@num_errors`.

If I add one more thing, the config name, `raise_error_on_batch_request_failure`, is about the implementation detail. It might be better if it explains higher-level behavior (e.g. `ignore_batch_request_failure_after_retry` with default value: true).

Comment (simukappu) [reply lib/fluent/plugin/kinesis_helper/api.rb]: Thank you for your feedback @jooohn, I totally agree with you about config name. I've renamed `raise_error_on_batch_request_failure` to `drop_failed_records_after_batch_request_retries` with default true.

What do you think about it?

Comment (jooohn) [reply lib/fluent/plugin/kinesis_helper/api.rb]: LGTM!

**Comment by simukappu** [comment]
We'll close this PR once and merge included commits into master branch. We'll publish this feature as v3.4.0.rc1. Thank you!
